File: apps/web/lib/smart-quote-api.ts
import { buildApiUrl } from "@/lib/api-base";

export type OptimizeMode = "PRICE" | "SPEED" | "MARGIN" | "PREMIUM" | "BALANCE";

export type SmartOffer = {
  id: number;
  clave: string | null;
  numParte: string | null;
  nombre: string | null;
  modelo: string | null;
  marca: string | null;
  categoria: string | null;
  subcategoria: string | null;
  descripcion: string | null;
  imagen: string | null;
  ean?: string | null;
  upc?: string | null;
  precio: number;
  moneda: string;
  tipoCambio: number | null;
  costMxn: number;
  stockTotal: number;
  stockPreferred: number;
  stockByWarehouse?: Array<{ code: string; qty: number; label?: string; city?: string }>;
  leadTimeDays: number;
  protegido: boolean;
  activo?: boolean;
  sustituto: string | null;
  especificaciones?: Array<{ tipo: string; valor: string }>;
  promociones?: unknown[];
  hasPromo?: boolean;
  score: number;
  badges: string[];
  sellPriceSuggested: number;
  marginPercent: number;
};

export type LogisticsZone = {
  id?: number;
  zoneCode: string;
  zoneName: string;
  baseCost: number | string;
  basePrice: number | string;
  active?: boolean;
};

export type QuoteLinePayload = {
  productCtId?: number;
  category?: string;
  name: string;
  description?: string | null;
  brand?: string | null;
  model?: string | null;
  sku?: string | null;
  partNumber?: string | null;
  unit?: string;
  qty: number;
  unitPrice: number;
  unitCost?: number | null;
  supplierSku?: string | null;
  supplierCode?: string | null;
  marginPercent?: number | null;
  stockSnapshot?: number | null;
  leadTimeDays?: number | null;
  scoreReason?: string | null;
  optimizationMode?: string | null;
  discount: number;
  tax: number;
  laborHours?: number;
  laborRate?: number;
  deliveryTime?: string | null;
};

async function sqRequest<T>(
  path: string,
  token: string,
  init?: RequestInit,
  fallback = "Error Smart Quote",
): Promise<T> {
  const res = await fetch(buildApiUrl(path), {
    ...init,
    headers: {
      Authorization: `Bearer ${token}`,
      ...(init?.body ? { "Content-Type": "application/json" } : {}),
      ...(init?.headers || {}),
    },
  });
  if (init?.signal?.aborted) {
    throw new DOMException("Aborted", "AbortError");
  }
  if (!res.ok) {
    const data = await res.json().catch(() => ({}));
    throw new Error(typeof data?.message === "string" ? data.message : fallback);
  }
  return res.json() as Promise<T>;
}

export async function smartQuoteSearch(
  token: string,
  params: {
    q?: string;
    brand?: string;
    category?: string;
    optimize?: OptimizeMode;
    targetMargin?: number;
    inStockOnly?: boolean;
    take?: number;
  },
  init?: RequestInit,
) {
  const qs = new URLSearchParams();
  if (params.q) qs.set("q", params.q);
  if (params.brand) qs.set("brand", params.brand);
  if (params.category) qs.set("category", params.category);
  if (params.optimize) qs.set("optimize", params.optimize);
  if (params.targetMargin != null) qs.set("targetMargin", String(params.targetMargin));
  if (params.inStockOnly === false) qs.set("inStockOnly", "0");
  if (params.take) qs.set("take", String(params.take));
  return sqRequest<{
    data: SmartOffer[];
    meta: {
      totalCandidates: number;
      mode: OptimizeMode;
      preferredWarehouse?: string;
      warehouses?: Array<{ code: string; label: string; city?: string }>;
    };
  }>(
    `smart-quote/search?${qs}`,
    token,
    init,
    "No se pudo buscar en catálogo CT",
  );
}

export async function smartQuoteCtStatus(token: string) {
  return sqRequest<{
    total: number;
    active: number;
    withStock: number;
    preferredWarehouse?: string;
    warehouses?: Array<{ code: string; label: string; city?: string }>;
    lastSync: { finishedAt?: string; status?: string; rowsUpserted?: number } | null;
  }>("smart-quote/ct/status", token, undefined, "No se pudo leer estado CT");
}

export async function smartQuoteFacets(token: string) {
  return sqRequest<{
    brands: Array<{ name: string | null; count: number }>;
    categories: Array<{ name: string | null; count: number }>;
  }>("smart-quote/facets", token, undefined, "No se pudieron cargar filtros CT");
}

export async function smartQuoteLogistics(token: string) {
  return sqRequest<LogisticsZone[]>(
    "smart-quote/logistics",
    token,
    undefined,
    "No se pudieron cargar zonas logísticas",
  );
}

/** Detalle de un SKU CT (vía búsqueda exacta hasta endpoint dedicado). */
export async function smartQuoteProduct(token: string, clave: string) {
  const res = await smartQuoteSearch(token, { q: clave, take: 8, inStockOnly: false });
  const key = clave.trim().toLowerCase();
  return res.data.find((o) => o.clave?.toLowerCase() === key) ?? res.data[0] ?? null;
}

export async function smartQuoteSubstitutes(
  token: string,
  clave: string,
  params?: { optimize?: OptimizeMode; targetMargin?: number; take?: number },
) {
  const qs = new URLSearchParams();
  if (params?.optimize) qs.set("optimize", params.optimize);
  if (params?.targetMargin != null) qs.set("targetMargin", String(params.targetMargin));
  if (params?.take) qs.set("take", String(params.take));
  const suffix = qs.toString() ? `?${qs}` : "";
  return sqRequest<SmartOffer[]>(
    `smart-quote/substitutes/${encodeURIComponent(clave)}${suffix}`,
    token,
    undefined,
    "No se pudieron cargar sustitutos",
  );
}

export async function smartQuoteTriggerSync(token: string, source: "PRIMARY" | "FULL" = "PRIMARY") {
  return sqRequest<unknown>(
    "smart-quote/ct/sync",
    token,
    { method: "POST", body: JSON.stringify({ source, async: false }) },
    "No se pudo sincronizar CT",
  );
}

export async function smartQuoteLaborSuggest(
  token: string,
  lines: Array<{ category?: string; qty: number; name?: string }>,
) {
  return sqRequest<
    Array<{
      code: string;
      name: string;
      category: string;
      qty: number;
      unitPrice: number;
      unitCost: number;
      laborHours: number;
      laborRate: number;
      reason: string;
    }>
  >(
    "smart-quote/labor/suggest",
    token,
    { method: "POST", body: JSON.stringify({ lines }) },
    "No se pudo sugerir mano de obra",
  );
}

export async function smartQuoteCheckMargin(
  token: string,
  body: {
    unitCost: number;
    unitPrice: number;
    discountPercent?: number;
    category?: string;
    brand?: string;
  },
) {
  return sqRequest<{
    ok: boolean;
    marginPercent: number;
    minRequired: number;
    requiresApproval: boolean;
    message: string | null;
  }>(
    "smart-quote/rules/check-margin",
    token,
    { method: "POST", body: JSON.stringify(body) },
    "No se pudo validar margen",
  );
}

export async function smartQuoteConfigure(
  token: string,
  body: {
    template: "CCTV" | "WIFI" | "ACCESS" | "CUSTOM";
    cameras?: number;
    storageDays?: number;
    accessPoints?: number;
    doors?: number;
    optimize?: OptimizeMode;
    targetMarginPercent?: number;
    logisticsZone?: string;
    includeLabor?: boolean;
  },
) {
  return sqRequest<{
    hardware: QuoteLinePayload[];
    labor: Array<{
      name: string;
      category: string;
      qty: number;
      unitPrice: number;
      unitCost: number;
      laborHours: number;
      laborRate: number;
      reason: string;
    }>;
    logistics: QuoteLinePayload | null;
    notes: string[];
  }>("smart-quote/configure", token, { method: "POST", body: JSON.stringify(body) }, "No se pudo configurar solución");
}

export async function smartQuoteCopilotDraft(token: string, prompt: string) {
  return sqRequest<{
    intent: { summary: string; questions: string[]; optimize: OptimizeMode; template: string };
    proposal: {
      hardware: QuoteLinePayload[];
      labor: Array<Record<string, unknown>>;
      logistics: QuoteLinePayload | null;
      notes: string[];
    };
    disclaimer: string;
  }>(
    "smart-quote/copilot/draft",
    token,
    { method: "POST", body: JSON.stringify({ prompt }) },
    "No se pudo generar borrador",
  );
}

/** IVA México. CT publica precio de lista sin IVA; se aplica al armar la cotización. */
export const CT_IVA_PERCENT = 16;

/** Precio de venta neto desde costo y margen % sobre venta. */
export function sellFromCost(cost: number, marginPercent: number) {
  const m = Math.min(0.95, Math.max(0.01, marginPercent / 100));
  return Math.round((cost / (1 - m)) * 100) / 100;
}

export function offerToLine(
  offer: SmartOffer,
  qty = 1,
  optimize: OptimizeMode = "BALANCE",
  targetMarginPercent?: number,
): QuoteLinePayload {
  const margin = targetMarginPercent ?? offer.marginPercent ?? 30;
  const unitCost = offer.costMxn;
  const unitPrice =
    targetMarginPercent != null ? sellFromCost(unitCost, margin) : offer.sellPriceSuggested;
  return {
    productCtId: offer.id,
    category: offer.categoria || "CT",
    name: offer.nombre || offer.clave || "Producto",
    description: offer.descripcion,
    brand: offer.marca,
    model: offer.modelo,
    sku: offer.clave,
    partNumber: offer.numParte,
    unit: "pieza",
    qty,
    unitPrice,
    unitCost,
    supplierSku: offer.clave,
    supplierCode: "CT",
    marginPercent: margin,
    stockSnapshot: offer.stockTotal,
    leadTimeDays: offer.leadTimeDays,
    scoreReason: offer.badges[0] || "RECOMMENDED",
    optimizationMode: optimize,
    discount: 0,
    tax: CT_IVA_PERCENT,
    deliveryTime: offer.leadTimeDays <= 1 ? "Inmediata" : `${offer.leadTimeDays} días`,
  };
}

export type CtEnvioForm = {
  nombre: string;
  direccion: string;
  entreCalles?: string;
  noExterior: string;
  noInterior?: string;
  colonia: string;
  estado: string;
  ciudad: string;
  codigoPostal: string;
  telefono: string;
};

export type CtOrderPreview = {
  lines: Array<{
    clave: string | null;
    nombre: string;
    qty: number;
    unitCost: number;
    unitSell: number;
    lineCost: number;
    lineSell: number;
    marginPercent: number | null;
  }>;
  subtotalCost: number;
  subtotalSell: number;
  marginAmount: number;
  quoteStatus: string;
  defaultEnvio: CtEnvioForm | null;
  config: { apiConfigured: boolean; defaultAlmacen: string; warehouses: Array<{ code: string; label: string }> };
  existingOrders: SupplierPurchaseOrderRow[];
};

export type SupplierPurchaseOrderRow = {
  id: number;
  status: string;
  externalFolio: string | null;
  almacen: string;
  errorMessage: string | null;
  createdAt: string;
  confirmedAt: string | null;
};

export type SupplierStatsResponse = {
  suppliers: Array<{
    supplierCode: string;
    label: string;
    lineCount: number;
    quoteCount: number;
    costNet: number;
    sellNet: number;
    taxAmount: number;
    sellWithTax: number;
    marginAmount: number;
    marginPercent: number;
    priceIncludesTax: boolean;
    customerTaxPercent: number;
  }>;
  totals: {
    quoteCount: number;
    costNet: number;
    sellNet: number;
    taxAmount: number;
    sellWithTax: number;
    marginAmount: number;
    marginPercent: number;
  };
};

export async function smartQuoteSupplierStats(
  token: string,
  params?: { from?: string; to?: string; status?: string },
) {
  const qs = new URLSearchParams();
  if (params?.from) qs.set("from", params.from);
  if (params?.to) qs.set("to", params.to);
  if (params?.status) qs.set("status", params.status);
  const suffix = qs.toString() ? `?${qs}` : "";
  return sqRequest<SupplierStatsResponse>(`smart-quote/supplier-stats${suffix}`, token);
}

export async function smartQuoteCtOrderPreview(token: string, cotizacionId: number) {
  return sqRequest<CtOrderPreview>(`smart-quote/ct/orders/preview/${cotizacionId}`, token);
}

export async function smartQuoteCtOrderSubmit(
  token: string,
  cotizacionId: number,
  body: {
    almacen: string;
    confirm?: boolean;
    envio: CtEnvioForm[];
  },
) {
  return sqRequest<SupplierPurchaseOrderRow>(
    `smart-quote/ct/orders/${cotizacionId}`,
    token,
    { method: "POST", body: JSON.stringify(body) },
    "No se pudo enviar pedido a CT",
  );
}

export async function smartQuoteCtOrderConfirm(token: string, orderId: number) {
  return sqRequest<SupplierPurchaseOrderRow>(
    `smart-quote/ct/orders/confirm/${orderId}`,
    token,
    { method: "POST", body: "{}" },
    "No se pudo confirmar pedido CT",
  );
}

File: apps/mobile-native/android/app/src/main/java/mx/nexara/mobile/nativeapp/data/api/SmartQuoteApi.kt
package mx.nexara.mobile.nativeapp.data.api

import com.squareup.moshi.Json
import retrofit2.http.Body
import retrofit2.http.GET
import retrofit2.http.POST
import retrofit2.http.Path
import retrofit2.http.Query

data class StockByWarehouseDto(
    val code: String = "",
    val qty: Int = 0,
    val label: String? = null,
    val city: String? = null,
)

data class SmartOfferDto(
    val id: Long = 0,
    val clave: String? = null,
    val numParte: String? = null,
    val nombre: String? = null,
    val modelo: String? = null,
    val marca: String? = null,
    val categoria: String? = null,
    val precio: Double = 0.0,
    val costMxn: Double = 0.0,
    val stockTotal: Int = 0,
    val stockPreferred: Int = 0,
    val stockByWarehouse: List<StockByWarehouseDto>? = null,
    val leadTimeDays: Int = 0,
    val promociones: List<Map<String, Any?>>? = null,
    val sellPriceSuggested: Double = 0.0,
    val marginPercent: Double = 0.0,
    val imagen: String? = null,
    val badges: List<String>? = null,
)

data class SmartSearchResponse(
    val data: List<SmartOfferDto> = emptyList(),
    val meta: SmartSearchMeta? = null,
)

data class SmartSearchMeta(
    val totalCandidates: Int = 0,
    val mode: String? = null,
)

data class SmartCtStatusDto(
    val total: Int = 0,
    val active: Int = 0,
    val withStock: Int = 0,
)

data class CreateCotizacionItemBody(
    val name: String,
    val qty: Int,
    val unitPrice: Double,
    val discount: Double = 0.0,
    val tax: Double = 16.0,
    val description: String? = null,
    val category: String? = null,
    val brand: String? = null,
    val model: String? = null,
    val sku: String? = null,
    val partNumber: String? = null,
    val unitCost: Double? = null,
    val supplierCode: String? = null,
    val productCtId: Long? = null,
    val supplierSku: String? = null,
    val marginPercent: Double? = null,
    val stockSnapshot: Int? = null,
    val leadTimeDays: Int? = null,
    val optimizationMode: String? = null,
)

data class CreateCotizacionBody(
    val quoteNumber: String,
    val issueDate: String,
    val validUntil: String,
    val clientName: String,
    val clientCompany: String? = null,
    val projectName: String? = null,
    val currency: String = "MXN",
    val status: String = "DRAFT",
    val items: List<CreateCotizacionItemBody>,
)

data class CreateCotizacionResponse(
    val id: Long? = null,
    @Json(name = "quoteNumber") val quoteNumber: String? = null,
)

data class SupplierStatRowDto(
    val supplierCode: String = "",
    val label: String = "",
    val lineCount: Int = 0,
    val quoteCount: Int = 0,
    val costNet: Double = 0.0,
    val sellNet: Double = 0.0,
    val taxAmount: Double = 0.0,
    val sellWithTax: Double = 0.0,
    val marginAmount: Double = 0.0,
    val marginPercent: Double = 0.0,
    val priceIncludesTax: Boolean = false,
    val customerTaxPercent: Double = 16.0,
)

data class SupplierStatsTotalsDto(
    val quoteCount: Int = 0,
    val costNet: Double = 0.0,
    val sellNet: Double = 0.0,
    val taxAmount: Double = 0.0,
    val sellWithTax: Double = 0.0,
    val marginAmount: Double = 0.0,
    val marginPercent: Double = 0.0,
)

data class SupplierStatsResponseDto(
    val suppliers: List<SupplierStatRowDto> = emptyList(),
    val totals: SupplierStatsTotalsDto? = null,
)

data class CtOrderLineDto(
    val clave: String? = null,
    val nombre: String = "",
    val qty: Int = 0,
    val unitCost: Double = 0.0,
    val unitSell: Double = 0.0,
    val lineCost: Double = 0.0,
    val lineSell: Double = 0.0,
    val marginPercent: Double? = null,
)

data class CtOrderPreviewDto(
    val lines: List<CtOrderLineDto> = emptyList(),
    val subtotalCost: Double = 0.0,
    val subtotalSell: Double = 0.0,
    val marginAmount: Double = 0.0,
    val quoteStatus: String = "",
    val existingOrders: List<CtSupplierOrderDto> = emptyList(),
)

data class CtSupplierOrderDto(
    val id: Long = 0L,
    val status: String = "",
    val externalFolio: String? = null,
)

data class CtOrderSubmitBody(
    val almacen: String,
    val confirmNow: Boolean = false,
)

data class SmartFacetDto(
    val name: String? = null,
    val count: Int = 0,
)

data class SmartFacetsResponseDto(
    val brands: List<SmartFacetDto> = emptyList(),
    val categories: List<SmartFacetDto> = emptyList(),
)

data class LaborSuggestItemDto(
    val code: String = "",
    val name: String = "",
    val category: String = "",
    val qty: Int = 1,
    val unitPrice: Double = 0.0,
    val unitCost: Double = 0.0,
    val laborHours: Double = 0.0,
    val laborRate: Double = 0.0,
    val reason: String = "",
)

data class MarginCheckDto(
    val ok: Boolean = true,
    val marginPercent: Double = 0.0,
    val minRequired: Double = 0.0,
    val requiresApproval: Boolean = false,
    val ruleName: String? = null,
    val message: String? = null,
)

data class CommercialRuleDto(
    val id: Int = 0,
    val name: String = "",
    val scope: String = "GLOBAL",
    val scopeValue: String? = null,
    val minMarginPercent: Double? = null,
    val maxDiscountPercent: Double? = null,
    val requiresApproval: Boolean = true,
    val active: Boolean = true,
)

data class LogisticsZoneDto(
    val zoneCode: String = "",
    val zoneName: String = "",
    val baseCost: Double = 0.0,
    val basePrice: Double = 0.0,
    val active: Boolean = true,
)

data class CopilotHardwareLineDto(
    val name: String = "",
    val qty: Int = 1,
    val unitPrice: Double = 0.0,
    val unitCost: Double = 0.0,
    val brand: String? = null,
    val sku: String? = null,
    val supplierCode: String? = null,
    val productCtId: Long? = null,
    val marginPercent: Double? = null,
)

data class CopilotProposalDto(
    val hardware: List<CopilotHardwareLineDto> = emptyList(),
    val notes: List<String> = emptyList(),
)

data class CopilotDraftDto(
    val proposal: CopilotProposalDto? = null,
    val disclaimer: String? = null,
)

data class LaborSuggestBody(
    val lines: List<Map<String, Any?>>,
)

data class MarginCheckBody(
    val unitCost: Double,
    val unitPrice: Double,
    val discountPercent: Double = 0.0,
    val category: String? = null,
    val brand: String? = null,
)

data class CopilotDraftBody(
    val prompt: String,
)

data class ConfigureSolutionBody(
    val template: String,
    val cameras: Int? = null,
    val storageDays: Int? = null,
    val accessPoints: Int? = null,
    val doors: Int? = null,
    val optimize: String = "BALANCE",
    val targetMarginPercent: Int = 30,
    val logisticsZone: String? = null,
    val includeLabor: Boolean = true,
)

data class ConfigureSolutionResponse(
    val hardware: List<CopilotHardwareLineDto> = emptyList(),
    val labor: List<LaborSuggestItemDto> = emptyList(),
    val logistics: CopilotHardwareLineDto? = null,
    val notes: List<String> = emptyList(),
)

interface SmartQuoteApi {
    @GET("smart-quote/search")
    suspend fun search(
        @Query("q") q: String? = null,
        @Query("brand") brand: String? = null,
        @Query("category") category: String? = null,
        @Query("optimize") optimize: String? = "BALANCE",
        @Query("targetMargin") targetMargin: Int? = 30,
        @Query("inStockOnly") inStockOnly: String? = "1",
        @Query("take") take: Int? = 40,
    ): SmartSearchResponse

    @GET("smart-quote/facets")
    suspend fun facets(): SmartFacetsResponseDto

    @GET("smart-quote/substitutes/{clave}")
    suspend fun substitutes(
        @Path("clave") clave: String,
        @Query("optimize") optimize: String? = "BALANCE",
        @Query("targetMargin") targetMargin: Int? = 30,
    ): List<SmartOfferDto>

    @POST("smart-quote/labor/suggest")
    suspend fun laborSuggest(@Body body: LaborSuggestBody): List<LaborSuggestItemDto>

    @GET("smart-quote/rules")
    suspend fun marginRules(): List<CommercialRuleDto>

    @GET("smart-quote/logistics")
    suspend fun logisticsZones(): List<LogisticsZoneDto>

    @POST("smart-quote/rules/check-margin")
    suspend fun checkMargin(@Body body: MarginCheckBody): MarginCheckDto

    @POST("smart-quote/copilot/draft")
    suspend fun copilotDraft(@Body body: CopilotDraftBody): CopilotDraftDto

    @POST("smart-quote/configure")
    suspend fun configure(@Body body: ConfigureSolutionBody): ConfigureSolutionResponse

    @GET("smart-quote/ct/status")
    suspend fun ctStatus(): SmartCtStatusDto

    @GET("smart-quote/supplier-stats")
    suspend fun supplierStats(
        @Query("from") from: String? = null,
        @Query("to") to: String? = null,
        @Query("status") status: String? = null,
    ): SupplierStatsResponseDto

    @GET("smart-quote/ct/orders/preview/{cotizacionId}")
    suspend fun ctOrderPreview(@Path("cotizacionId") cotizacionId: Long): CtOrderPreviewDto

    @POST("smart-quote/ct/orders/{cotizacionId}")
    suspend fun ctOrderSubmit(
        @Path("cotizacionId") cotizacionId: Long,
        @Body body: CtOrderSubmitBody,
    ): Map<String, Any?>

    @POST("smart-quote/ct/orders/confirm/{orderId}")
    suspend fun ctOrderConfirm(
        @Path("orderId") orderId: Long,
    ): Map<String, Any?>

    @POST("cotizaciones")
    suspend fun createCotizacion(@Body body: CreateCotizacionBody): CreateCotizacionResponse
}
